# Escuela Politécnica Nacional

## Métodos Numericos
## Isaac Gonzalez
## 18/10/2025
---

# Tarea 2

Resuelva los siguientes ejercicios, tome en cuenta que debe mostrar el desarollo completo del ejercicio.

1. Calcule los errores absoluto y relativo en las aproximaciones de $p$ por $p^*$

| Error Absoluto | Error Relativo |
|----------------|----------------|
| $\lvert p - p^* \rvert$ | $\left\lvert \dfrac{p - p^*}{p} \right\rvert$ |



#### a. $p = \pi $ , $ p^* = \frac{22}{7} $

Error Abs:

$ = \lvert \pi - 22/7 \rvert$  $ = 1.264489267\times 10^{-3} $

Error Rel:

$ = \left\lvert \dfrac{\pi - 22/7}{\pi} \right\rvert$  $ = 4.024994348\times 10^{-4}$

#### b. $ p = \pi $, $ p^* = 3.1416 $

Error Abs:

$ = \lvert \pi - 3.1416 \rvert$  $ = 7.3464102\times 10^{-6} $

Error Rel:

$ = \left\lvert \dfrac{\pi - 3.1416}{\pi} \right\rvert$  $ = 2.338434995\times 10^{-6}$

#### c. $ p = e $,  $ p^* = 2.718 $

Error Abs:

$ = \lvert e - 2.718 \rvert$  $ = 2.81828459\times 10^{-4} $

Error Rel:

$ = \left\lvert \dfrac{ e - 2.718}{e} \right\rvert$  $ = 1.03678896\times 10^{-4}$

#### d. $ p = \sqrt{2} $,  $ p^* = 1.414 $

Error Abs:

$ = \lvert \sqrt{2} - 1.414 \rvert$  $ = 2.135623731\times 10^{-4} $

Error Rel:

$ = \left\lvert \dfrac{ \sqrt{2} - 1.414}{\sqrt{2}} \right\rvert$  $ = 1.510114022\times 10^{-4}$

In [6]:
import math
from math import sqrt
def calcular_error_absoluto(p_real, p_apr):
    return abs(p_real - p_apr)

def calcular_error_relativo(p_real, p_apr):
    if p_real == 0:
        raise ValueError("El valor real no puede ser cero para calcular el error relativo.")
    return abs((p_real - p_apr) / p_real)  

def main():
    p_real = math.factorial(9)
    p_apr = math.sqrt(18 * math.pi) * (9 / math.e) ** 9 
    error_abs = calcular_error_absoluto(p_real, p_apr)
    error_rel = calcular_error_relativo(p_real, p_apr)
    print(f"El error_abs: " + str(calcular_error_absoluto(p_real, p_apr)))
    print(f"El error_rel: " + str(calcular_error_relativo(p_real, p_apr)))
    
if __name__ == "__main__":
    main() 

El error_abs: 3343.1271580516477
El error_rel: 0.009212762230080598


---

2. Calcule los errores absoluto y relativo en las aproximaciones de $p$ por $p^*$

##### a. $ p = e^{10} $,  $ p^* = 22000 $

Error Abs:

$ = \lvert \sqrt{2} - 1.414 \rvert$  $ = 2.135623731\times 10^{-4} $

Error Rel:

$ = \left\lvert \dfrac{ \sqrt{2} - 1.414}{\sqrt{2}} \right\rvert$  $ = 1.510114022\times 10^{-4}$

##### b. $ p = 10^{\pi} $,  $ p^* = 1400 $

Error Abs:

$ = \lvert 10^{\pi} - 1400 \rvert$  $ = 14.54426863 $

Error Rel:

$ = \left\lvert \dfrac{ 10^{\pi} - 1400}{10^{\pi}} \right\rvert$  $ = 0.0104978227$

##### c. $ p = 8! $,  $ p^* = 39900 $

Error Abs:

$ = \lvert 8! - 39900 \rvert$  $ = 420 $

Error Rel:

$ = \left\lvert \dfrac{ 8! - 39900}{8!} \right\rvert$  $ = 0.01041666667$

##### d. $ p = 9! $,  $ p^* = \sqrt{18\pi}\times (\frac{9}{e})^9 $

Error Abs:

$ = \lvert 9! - \sqrt{18\pi}\times (\frac{9}{e})^9 \rvert$  $ = 3343,127158 $

Error Rel:

$ = \left\lvert \dfrac{ 9! - \sqrt{18\pi}\times (\frac{9}{e})^9}{9!} \right\rvert$  $ = 9.21276223\times 10^{-13}$


---

3. Encuentre el intervalo más largo en el que se debe encontrar $p^*$ para aproximarse a $p$ con error relativo máximo de $10^{-4}$ para cada valor $p$



In [3]:
import math

def intervalo_aproximacion(p, error_rel_max=1e-4):
    if p == 0:
        return (-error_rel_max, error_rel_max)
    else:
        delta = abs(p) * error_rel_max
        return (p - delta, p + delta)
    

#  1: p = π
p = math.pi
intervalo_pi = intervalo_aproximacion(p)
print(f"Para p = π ≈ {p}")
print(f"Intervalo: [{intervalo_pi[0]:.10f}, {intervalo_pi[1]:.10f}]")

#Ejemplo 2: p = e
p = math.e
intervalo_e = intervalo_aproximacion(p)
print(f"\nPara p = e ≈ {p}")
print(f"Intervalo: [{intervalo_e[0]:.10f}, {intervalo_e[1]:.10f}]")

# 3: p = sqrt{2}

p = math.sqrt(2)
intervalo_sqr = intervalo_aproximacion(p)
print(f"\nPara p = sqrt(2) ≈ {p} ")
print(f"Intervalo: [{intervalo_sqr[0]:.10f}, {intervalo_sqr[1]:.10f}]")

# 4: p = raiz cubica de (7)
p = 7 ** (1/3)
intervalo_raiz7 = intervalo_aproximacion(p)
print(f"\nPara p = raíz cubica de 7 ≈ {p}")
print(f"Intervalo: [{intervalo_raiz7[0]:.10f}, {intervalo_raiz7[1]:.10f}]")

    

Para p = π ≈ 3.141592653589793
Intervalo: [3.1412784943, 3.1419068129]

Para p = e ≈ 2.718281828459045
Intervalo: [2.7180100003, 2.7185536566]

Para p = sqrt(2) ≈ 1.4142135623730951 
Intervalo: [1.4140721410, 1.4143549837]

Para p = raíz cubica de 7 ≈ 1.912931182772389
Intervalo: [1.9127398897, 1.9131224759]



* Procedimiento general:

$\frac{|p - p^*|}{|p|}$ $ \leq 10^{-4} $

* Se despeja.

$|p-p^*| \leq 10^{-4} \times |p| $

* Lo que signfíca que $p^*$ va estar en el intervalo:

$p^* \in [p - 10^{-4}\times |p| , p + 10^{-4}\times |p|] $

$p > 0  $   and  $ p < 0 $


#### a. $\pi$

* $ p = \pi $
* $\Delta = 10^{-4} \times |p| \approx 10^{-4} \times |\pi| $ 
* Intervalo: $[3.1412784943, 3.1419068129]$

#### b. $e$

* $ p = e $
* $\Delta = 10^{-4} \times |p| \approx 10^{-4} \times |e| $ 
* Intervalo: $[2.7180100003, 2.7185536566]$

#### c. $\sqrt{2}$

* $ p = \sqrt{2} $
* $\Delta = 10^{-4} \times |p| \approx 10^{-4} \times |\sqrt{2}| $ 
* Intervalo: $[1.4140721410, 1.4143549837]$

#### d. $\sqrt[3]{7}$

* $ p = \sqrt[3]{7} $
* $\Delta = 10^{-4} \times |p| \approx 10^{-4} \times |\sqrt[3]{7}| $ 
* Intervalo: $[1.9127398897, 1.9131224759]$

---

4. Use la aritmética de redondeo de tres dígitos para realizar lo siguiente. Calcule los errores absoluto y relativo
con el valor exacto determinado para por lo menos cinco dígitos.

#### a. $\frac{\frac{13}{14}-\frac{5}{7}}{2e-5.4}  $
* $\frac{13}{14} = 0.928571 \approx 0.929 $
* $\frac{5}{7} = 0.714285 \approx 0.714 $
* Numerador $= 0.929 - 0.714 = 0.215 \approx 0.215$
* $2e = 5.43656 \approx 5.44 $
* Denominador $= 5.44 - 5.4 = 0.04 \approx 0.04$
* Cociente $= \frac{0.215}{0.04} = 5.375 \approx 5.38 $

Valor Exacto: p = 5.860620418

Errores:
* $ Error_{abs} = 0.4806204179 $
* $ Error_{rela} \approx 8.2008 \times 10^{-2}$

#### b. $-10\pi + 6e - \frac{3}{61}$

* $-10\pi = -31.4159 \approx -31.4 $
* $6e = 16.3097 \approx 16.3$
* Suma $= -31.4 + 16.3 = -15.1 \approx -15.1 $
* $\frac{3}{61} = 0.0491803 \approx 0.0492$
* resta $= -15.1 - 0.0492 = -15.1492 \approx -15.1$

Valor Exacto: p = -15.1554158930

Errores:
* $ Error_{abs} =  5.5416\times10^{-2}$
* $ Error_{rela} \approx 3.6565 \times 10^{-3}$

#### c. $(\frac{2}{9}) \times (\frac{9}{11}) $
 
 * $ \frac{2}{9} = 0.2222222 \approx 0.222 $
 * $ \frac{9}{11} = 0.8181818182 \approx 0.818$
 * Producto $= 0.222 \times 0.818 = 0.181996 \approx 0.182$

 Valor Exacto: p = 0.1818181818

Errores:
* $ Error_{abs} =  1.8182\times10^{-4}$
* $ Error_{rela} \approx 1.00000\times 10^{-3}$

#### d. $\frac{\sqrt{13} + \sqrt{11}}{\sqrt{13} - \sqrt{11}} $

* $ \sqrt{13} = 3.60555 \approx 3.61 $
* $ \sqrt{11} = 3.31662 \approx 3.32 $
* Numerador $= 3.61 + 3.32 = 6.93 \approx 6.93$
* Denominador $= 3.61 - 3.32 = 0.29 \approx 0.29$
* Cociente $= \frac{6.93}{0.29} = 23.8966 \approx 23.9$

 Valor Exacto: p = 23.9582607431

Errores:
* $ Error_{abs} =  5.8261\times10^{-2}$
* $ Error_{rela} \approx 2.4318\times 10^{-3}$


In [4]:
import math

def round_sig(x, sig=3):
    if x == 0: 
        return 0.0
    return round(x, sig - 1 - int(math.floor(math.log10(abs(x)))))

def b_three_digit():
    t1 = round_sig(-10*math.pi)
    t2 = round_sig(6*math.e)
    s = round_sig(t1 + t2)
    r = round_sig(3/61)
    val = round_sig(s - r)
    return val

def errors(approx, exact):
    ea = abs(approx - exact)
    er = ea/abs(exact) if exact != 0 else float("inf")
    return ea, er


B_ap, B_ex = b_three_digit(), -10*math.pi + 6*math.e - 3/61

for name, ap, ex in [('b',B_ap,B_ex)]:
    ea, er = errors(ap, ex)
    print(f"{name}) aprox={ap}, exacto={ex:.12f},  E_abs={ea:.12g},  E_rel={er:.12g}")

b) aprox=-15.1, exacto=-15.155415893013,  E_abs=0.0554158930125,  E_rel=0.00365650757483


---

5. 